What is the deal with security and distributed systems? (#306)
mauroservienti/milestone.github.io: _posts/security-and-soa.md

---
layout: post
title: "What is the deal with security and distributed systems?"
author: Mauro Servienti
synopsis: "Security is a crucial topic for any architect. We cannot implement it as a second thought. We must consider its implications from day one. Distributed systems are no different. However, it might be a little more involved."
enable_mermaid: true
header_image: /img/posts/security-and-soa/header.jpg
tags:
- soa
- distributed-systems
- security
---

If a system is composed of [autonomous components](https://milestone.topics.it/2022/09/05/autonomy.html), how can we guarantee the consistency imposed by security requirements?

For example, one common requirement might be formulated as follows: whenever an account is disabled or terminated, system access must be immediately blocked.

That usually raises the question, "if components are autonomous, how can the Security service impose behaviors without generating coupling?"

My answer usually starts with "Security is not a service."

## Wait, what?

> Security, as a term, is too generic in this context. At least, we want to distinguish between authentication, authorization, and users & rights management. Here is a quick recap of what they are:

Reviewer: Is this meant to be a bulleted list?

Author: Ah yes, I screwed markdown formatting.

>  Authentication verifies who's trying to access the system is who they claim to be. For example, credentials verification is a form of authentication. Another one is biometric checks.

Reviewer: ```suggestion
 Authentication verifies that the account trying to access the system is who they claim to be. For example, credentials verification is a form of authentication. Another one is biometric checks.
```

Authorization verifies that users can do what they are trying to do. For example, can the current user visualize a specific resource?

> Users & Rights management is the set of processes related to managing user accounts and what they can do in the system—for example, roles assignment.

Reviewer: ```suggestion
Users & Rights management is the set of processes related to managing user accounts and what they can do in the system — for example, roles assignment.
```

> Users & rights management is a good candidate for being a service. It's a set of processes well isolated from other services. It comes with its policies, and unlikely there is a need to couple it with other components in the system.

Reviewer: ```suggestion
Users & rights management is a good candidate for being a service. It's a set of processes well isolated from other services. It comes with its policies, and it's unlikely there is a need to couple it with other components in the system.
```

> Authentication is two-fold. On the one end, it is one of the Users & Rights management components. Unauthenticated users accessing the system will be redirected to the authentication component to validate their credentials. The component will query the Users & Rights management storage to check credentials or use the configured technology to authenticate them. Upon successful validation, users receive an authentication token they can spend when accessing other services.

Reviewer: ```suggestion
Authentication is two-fold. On the one hand, it is one of the Users & Rights management components. Unauthenticated users accessing the system will be redirected to the authentication component to validate their credentials. The component will query the Users & Rights management storage to check credentials or use the configured technology to authenticate them. Upon successful validation, users receive an authentication token they can spend when accessing other services.
```

## Is that an issue?

You bet it is. We now have a single point of failure. But again, it's primarily a business-caused issue. The business requirements force us in that direction. There is only so much we can do about it other than focusing on scaling out the active components and doing our best to provide high availability. In the end, it's what the business wants. A requirement such as the above states that the business favors a non-functioning system over a potentially insecure one.

## It's not all, folks: tokens expire!

So far, it went without saying that authentication and authorization happen at the system's outer boundaries. For example, web application front-ends validate authentication tokens and verify Roles. That implies a message sent on a queue by a front-end to a back-end service flies into the system unauthenticated.

That is primarily because security tokens expire (for good reasons). For example, the Kerberos token lifetime is five minutes (IIRC). Active directory users need a new, freshly generated token every five minutes.

> Token expiration works in favor and against us at the same time. A token lifetime is a few minutes or more, which makes it behave like a cache. It helps in reducing the single point of failure issue highlighted above. If requirements allow, we can rely on the token "cache" to avoid going back to authentication and authorization components every time.

Reviewer: ```suggestion
Token expiration works for and against us at the same time. A token lifetime is a few minutes or more, which makes it behave like a cache. It helps in reducing the single point of failure issue highlighted above. If requirements allow, we can rely on the token "cache" to avoid going back to authentication and authorization components every time.
```
